THREESCALE-8997 Allow defining the pod priority of components via APIManager CR

valerymo opened the pull request:
## WHAT

Jira: https://issues.redhat.com/browse/THREESCALE-8997
Allow defining the pod priority of components via APIManager CR

**Added optional priorityClassName definition for following 3scale pods:** (testing WIP)
| Component  | Fresh install adds `priorityClassName` | Reconcile `priorityClassName` |
| ------------- | ------------- | ------------- |
| apicast-production |  [X]  |  [X]  |
| apicast-staging | [X]  |  [X]  |
|  backend-cron | [X]  | [X]  |
| backend-listener |  [X]  |  [X] |
| backend-worker |  [X] | [X]  |
| backend-redis |  [X]  | [X]  |
| system-app | [X] | [X]  |
| system-sidekiq |  [X]  |  [X]  |
| ~system-sphinx~ | [X] |  [X]  |
| system-searchd | [X] |  [X]  |
| system-memcache |  [X] |  [X]  |
| system-mysql |  [X] |  [X] |
| system-postgresql | [X] |  [X]  |
| system-redis |  [X]  |  [X]  |
| zync |  [X]  | [X]  |
| zync-database |  [X]  | [X]  |
| zync-que |  [X]  |  [X]  |

_ApiManager CR example for apicast-staging (see Validation section for CR example for all/other DCs/pods_
```yaml
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: example-apimanager
spec:
  wildcardDomain: <wildcardDomain>
  apicast:
    stagingSpec:
      priorityClassName: system-node-critical
```

_S3 secret example_
```yaml
kind: Secret
apiVersion: v1
metadata: 
  name: s3-credentials
data: 
  AWS_ACCESS_KEY_ID: QUtXXXXXXXXXXXX=
  AWS_SECRET_ACCESS_KEY: bmZl=
  AWS_BUCKET: dGV=
  AWS_REGION: dXMt==
type: Opaque
```

## Validation

#### Install 3scale operator

```
$ cd 3scale-operator
$ make install
```

```
$ export NAMESPACE=3scale-test
$ oc new-project $NAMESPACE
$ make download
$ make run
```
*Secret and CR to be installed on separate terminal window*. 

#### Secrets & CRs for test

- Secret's AWS fields - encoded base64:    `echo ABC |base64` 

```yaml
oc apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata: 
  name: s3-credentials
data: 
  AWS_ACCESS_KEY_ID: QUtJQVYXXXXXX=
  AWS_SECRET_ACCESS_KEY: bmZX=
  AWS_BUCKET: dGX=
  AWS_REGION: dXX==
type: Opaque
EOF
```

- Sample of CR for test 3scale pods

```yaml
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
    name: example-apimanager
spec:
    wildcardDomain: api.vmogilev01.0nno.s1.devshift.org
    resourceRequirementsEnabled: false
    apicast:
        stagingSpec:
            priorityClassName: openshift-user-critical
        productionSpec:
            priorityClassName: openshift-user-critical
    backend:
        listenerSpec:
            priorityClassName: openshift-user-critical
        cronSpec:
            priorityClassName: openshift-user-critical
        workerSpec:
            priorityClassName: openshift-user-critical
        redisPriorityClassName: openshift-user-critical
    system:
        appSpec:
            priorityClassName: openshift-user-critical
        sidekiqSpec:
            priorityClassName: openshift-user-critical
        sphinxSpec:
            priorityClassName: openshift-user-critical
        searchdSpec:
            priorityClassName: openshift-user-critical
        memcachedPriorityClassName: openshift-user-critical
        redisPriorityClassName: openshift-user-critical
        database:
            postgresql:
                priorityClassName: openshift-user-critical                
    zync:
        appSpec:
            priorityClassName: openshift-user-critical
        queSpec:
            priorityClassName: openshift-user-critical
        databasePriorityClassName: openshift-user-critical

```

Describe pods and see that  priorityClassName set
```
$ oc describe pod apicast-staging-1-xxxx |grep "Priority Class Name"
Priority Class Name:  system-node-critical

oc describe pod system-memcache-1-xxxx |grep "Priority Class Name"
Priority Class Name:  system-node-critical

$ oc describe pod backend-redis-2-xxxx |grep "Priority Class Name"
Priority Class Name:  system-node-critical
```
Check similar for other pods
```bash
kubectl get deploymentconfigs -o name | xargs -I {} bash -c "echo -n '{} -> priorityClassName: '; kubectl get {} -o jsonpath='{.spec.template.spec.priorityClassName}' | yq e -P"
```

MStokluska commented on pkg/3scale/amp/operator/system_reconciler.go:
```diff
@@ -95,7 +95,7 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
 	}
 
-	err = r.ReconcileDeploymentConfig(system.AppDeploymentConfig(), reconcilers.DeploymentConfigMutator(systemAppDCMutators...))
+	err = r.ReconcileDeploymentConfig(system.AppDeploymentConfig(*r.apiManager.Spec.PriorityClassName), reconcilers.DeploymentConfigMutator(systemAppDCMutators...))
```
@valerymo if the priorityClassName is optional field are we in risk here to get a nil pointer?

valerymo replied:
@MStokluska , thank you for comment. The approach/implementation changed. Now it's done via Options, as @eguzki  recommended.

MStokluska commented on pkg/3scale/amp/operator/system_reconciler.go:
```diff
@@ -113,7 +113,7 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
 	}
 
-	err = r.ReconcileDeploymentConfig(system.SidekiqDeploymentConfig(), reconcilers.DeploymentConfigMutator(sidekiqDCMutators...))
+	err = r.ReconcileDeploymentConfig(system.SidekiqDeploymentConfig(*r.apiManager.Spec.PriorityClassName), reconcilers.DeploymentConfigMutator(sidekiqDCMutators...))
```
same q as above

valerymo replied:
Thank you for comment, please see reply above

MStokluska commented on pkg/3scale/amp/operator/system_reconciler.go:
```diff
@@ -127,7 +127,7 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
 		upgrade.SphinxSecretKeyEnvVarMutator,
 	)
-	err = r.ReconcileDeploymentConfig(system.SphinxDeploymentConfig(), sphinxDCmutator)
+	err = r.ReconcileDeploymentConfig(system.SphinxDeploymentConfig(*r.apiManager.Spec.PriorityClassName), sphinxDCmutator)
```
same q as above

valerymo replied:
Thank you for comment, please see reply above

MStokluska commented on pkg/3scale/amp/component/system.go:
```diff
@@ -1215,7 +1216,7 @@ func (system *System) getSystemServiceDiscoveryData() string {
 }
 
-func (system *System) SphinxDeploymentConfig() *appsv1.DeploymentConfig {
+func (system *System) SphinxDeploymentConfig(priorityClassName string) *appsv1.DeploymentConfig {
```
are we missing priorityclassname below?

valerymo replied:
Thank you for comment, please see reply above

MStokluska commented:
In my opinion, I would pull the priorityClassName from within the each deployment config functions rather than from reconciler. Is systemOptions something we could leverage here to stay consistent?

eguzki commented on pkg/3scale/amp/component/system.go:
```diff
@@ -555,7 +554,7 @@ func (system *System) volumeNamesForSystemAppPreHookPod() []string {
 	return res
 }
 
-func (system *System) AppDeploymentConfig() *appsv1.DeploymentConfig {
```
`priorityClassName` should be passed in Options struct. The usual workflow is: 1) read by the options_provider and 2) populate the options struct.

valerymo replied:
@eguzki , thank you for comment and solution, I tried to do it, meanwhile for 3 components - system, backup and apicast; will continue,

valerymo replied:
Done

eguzki commented on apis/apps/v1alpha1/apimanager_types.go:
```diff
@@ -84,6 +84,8 @@ type APIManagerSpec struct {
 	// +optional
 	Monitoring *MonitoringSpec `json:"monitoring,omitempty"`
+	// +optional
+	PriorityClassName *string `json:"priorityClassName,omitempty"`
```
on global configuration for all? What if a customer wants to grant priority to some components only??

valerymo replied:
@eguzki , thank you for comment. Now I'm trying to add separate Priority classes for groups, but maybe need for each pod ... Could you please help , how better implement it. Currently i did it as following, but not sure it's good enough - https://github.com/3scale/3scale-operator/blob/e4a4eb54ae29f3715eb3d99efc8a0d18a69d6476/apis/apps/v1alpha1/apimanager_types.go#L88-L92

valerymo replied:
@eguzki This is new commit - priorityClass per Pod ; example for system - just to ask if it's ok , if yes, I will continue with other pods  -  https://github.com/3scale/3scale-operator/pull/808/commits/f42c9a193313dbdb338c5ca64cd7e7a7d5e38128 . Thank you

valerymo replied:
@eguzki , thank you for comment. I'm trying to do it for each pod now. Could you please look at latest commits. Thank you. 
cc @MStokluska

valerymo commented:
Hello @eguzki , @MStokluska , I'd like provide this PR for Review. I hope it's implemented now, as Eguzki suggested. I will be in PTO from tomorrow, as can't address your comments, but any notes will be appreciated and I can address it at the end of next week.
I'd like to note, that I still trying to test it, as seems some configuration or my environment issue (I hope), as I can't see that Priority changed for pod `apicast-staging`. 
Thank you for your review and help

valerymo commented:
Hello @eguzki  , @MStokluska , just wanted say that it's working for me now (I found annoying issue wher "PriorityClass" had typo (and it was reason of issue). Now - it's tested for apicast-staging (as appears in description validation notes). I will try do more tests, but wanted to ask ask you to review if you will have time.. Thank you

eguzki commented on pkg/3scale/amp/component/system.go:
```diff
@@ -2,6 +2,7 @@ package component
 
 import (
 	"fmt"
+	"github.com/3scale/3scale-operator/pkg/helper"
```
nit: :thinking:  why?

valerymo replied:
updated, thank you

eguzki replied:
I do not see it updated

I am missing something?

valerymo replied:
hope it's updated now. Thank you

eguzki commented on pkg/3scale/amp/operator/redis_options_provider.go:
```diff
@@ -242,3 +243,9 @@ func (r *RedisOptionsProvider) backendRedisPodTemplateLabels() map[string]string
 	return labels
 }
+
+func (r *RedisOptionsProvider) setPriorityClassNames() {
```
Missing backend redis priority class name ?

valerymo replied:
Added. Thank you

eguzki commented on apis/apps/v1alpha1/apimanager_types.go:
```diff
@@ -407,6 +419,8 @@ type SystemSpec struct {
 	// +optional
 	RedisResources *v1.ResourceRequirements `json:"redisResources,omitempty"`
+	// +optional
+	RedisPriorityClassName *string `json:"redisPriorityClassName,omitempty"`
```
missing the same one for backend redis?

valerymo replied:
Added. Thank you

eguzki commented:
Upgrade procedures were missing. Implemented, as an example, the upgrade for the apicast-staging DC. The others should be done in a similar way https://github.com/valerymo/3scale-operator/pull/3

valerymo commented on apis/apps/v1alpha1/apimanager_types.go:
```diff
@@ -464,6 +482,8 @@ type SystemSphinxSpec struct {
 	// +optional
 	Resources *v1.ResourceRequirements `json:"resources,omitempty"`
+	// +optional
+	PriorityClassName *string `json:"priorityClassName,omitempty"`
```
ok, thank you

valerymo commented:
Hi @eguzki . Priority Class Name upgrade -  implemented. Thank you very much for example.

eguzki commented:
/test test-e2e

eguzki commented on apis/apps/v1alpha1/apimanager_types.go:
```diff
@@ -597,6 +621,9 @@ type ZyncSpec struct {
 	QueSpec *ZyncQueSpec `json:"queSpec,omitempty"`
+
+	// +optional
+	DatabasePriorityClassName *string `json:"databasePriorityClassName,omitempty"`
```
Zync database priority class name is missing

eguzki reviewed (changes requested):
Missing upgrade for:
* backend-cron
* backend-listener
* backend-worker

eguzki commented:
Update PR description

*     system_postgresql -> should be implemented in this PR.
*    apimanager_backup -> this is not a deployment, hence it does not have priority class name

eguzki commented on pkg/3scale/amp/component/redis_options.go:
```diff
@@ -41,6 +44,8 @@ type RedisOptions struct {
 	SystemRedisSentinelsRole         string
 	SystemRedisNamespace             string
+
+	PriorityClassName string `validate:"-"`
```
leftover?

valerymo replied:
done. Thank you!

eguzki commented on pkg/3scale/amp/operator/system_mysql_options_provider.go:
```diff
@@ -202,3 +203,11 @@ func (s *SystemMysqlOptionsProvider) podTemplateLabels() map[string]string {
+
+func (s *SystemMysqlOptionsProvider) setPriorityClassNames() {
+	if s.apimanager.Spec.System.DatabaseSpec != nil &&
+		s.apimanager.Spec.System.DatabaseSpec.MySQL != nil &&
```
`s.apimanager.Spec.System.DatabaseSpec.MySQL` is not necessarily not nil to have Mysql deployed, as Mysql is the default. 

It is better to call `apìmanager.IsSystemMysqlEnabled()`

eguzki replied:
priorityClassname is not set for the system-mysql component

valerymo replied:
update done to use IsSystemMysqlEnabled.
Regarding "priorityClassname is not set for the system-mysql component" - it's implemented, but I will retest it. Thank you for comments

eguzki replied:
my bad, sorry. When `SystemMySQLReconciler` is called, it can safely be assumed mysql is enabled. 

I think it was right as it was

```go
if s.apimanager.Spec.System.DatabaseSpec != nil &&
		s.apimanager.Spec.System.DatabaseSpec.MySQL != nil &&
                s.apimanager.Spec.System.DatabaseSpec.MySQL.PriorityClassName != nil {
                s.mysqlOptions.PriorityClassName = *s.apimanager.Spec.System.DatabaseSpec.MySQL.PriorityClassName
}
```

valerymo commented:
> Missing upgrade for:
> 
> * backend-cron
> * backend-listener
> * backend-worker

@eguzki , sorry for delay with addressing the comments (PTO + other ...).  DeploymentConfigPriorityClassMutator was added to GenericBackendMutators, it is not right?

valerymo commented:
Hello @eguzki , hope I addressed all your comments. Thank you very much for your comments and help! Hope the testing table now could be completed. Could you please look when will have time. Thank you!

eguzki commented:
`system-sphinx` component has been deprecated (not dropped for now). The replacement is `system-searchd`. Priority class name should be implemented for fresh install and also upgrade. Upgrade is not strictly needed, but provides reconcile feature, so if the APImanager spec is changed, the deployment will also change.

eguzki commented on pkg/3scale/amp/operator/system_postgresql_options_provider.go:
```diff
@@ -189,3 +189,10 @@ func (s *SystemPostgresqlOptionsProvider) podTemplateLabels() map[string]string
 	return labels
 }
+
+func (s *SystemPostgresqlOptionsProvider) setPriorityClassNames() {
```
this method is never called. That's the why the postgresql pod does not get the priority class name assigned in the APIManager CR

valerymo replied:
done. thank you

eguzki commented:
`backend-redis` does not get the priority class name spec from the APIManager in fresh install mode

valerymo commented:
> `system-sphinx` component has been deprecated (not dropped for now). The replacement is `system-searchd`. Priority class name should be implemented for fresh install and also upgrade. Upgrade is not strictly needed, but provides reconcile feature, so if the APImanager spec is changed, the deployment will also change.

It's done, but problem found in Upgrade - pod rolling update is not working. Please see log bellow
```
$ oc describe pod system-searchd-1-d4t2t |grep Prio
Priority:             2000001000
Priority Class Name:  system-node-critical

$ oc edit APIManager example-apimanager
apimanager.apps.3scale.net/example-apimanager edited

$ oc get pod |grep searchd
system-searchd-1-d4t2t        1/1     Running             0               4m59s
system-searchd-1-deploy       0/1     Completed           0               5m
system-searchd-2-8bgpm        0/1     ContainerCreating   0               85s
system-searchd-2-deploy       1/1     Running             0               87s


**** Problem with searchd pod rolling update:
$ oc describe pod system-searchd-2-8bgpm
  Warning  FailedAttachVolume  11m                    attachdetach-controller  Multi-Attach error for volume "pvc-62ec0cb3-91ea-4d43-91c5-c09dd80143fa" Volume is already used by pod(s) system-searchd-1-d4t2t
  Warning  FailedMount         3m10s (x2 over 5m13s)  kubelet                  Unable to attach or mount volumes: unmounted volumes=[system-searchd-database], unattached volumes=[kube-api-access-cccgf system-searchd-database]: timed out waiting for the condition


```

valerymo commented:
> `backend-redis` does not get the priority class name spec from the APIManager in fresh install mode

Fixed. Thank you!
```
$ kubectl get deploymentconfigs -o name | xargs -I {} bash -c "echo -n '{} -> priorityClassName: '; kubectl get {} -o jsonpath='{.spec.template.spec.priorityClassName}' | yq e -P"
deploymentconfig.apps.openshift.io/apicast-production -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/apicast-staging -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-cron -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-listener -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-redis -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-worker -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-app -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-memcache -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-mysql -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-redis -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-searchd -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-sidekiq -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/zync -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/zync-database -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/zync-que -> priorityClassName: system-node-critical
)$ 
$ date
Thu May  4 13:55:11 IDT 2023
$ oc whoami --show-server
https://api.vmogilev01.0nno.s1.devshift.org:6443
```

valerymo commented:
@eguzki 
system-postgresql issue fixed
```
$ kubectl get deploymentconfigs -o name | xargs -I {} bash -c "echo -n '{} -> priorityClassName: '; kubectl get {} -o jsonpath='{.spec.template.spec.priorityClassName}' | yq e -P"
deploymentconfig.apps.openshift.io/apicast-production -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/apicast-staging -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-cron -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-listener -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-redis -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/backend-worker -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-app -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-memcache -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-postgresql -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-redis -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-searchd -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/system-sidekiq -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/zync -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/zync-database -> priorityClassName: system-node-critical
deploymentconfig.apps.openshift.io/zync-que -> priorityClassName: system-node-critical
```

After change Priority Class:
$ oc describe dc system-postgresql |grep -i prio
  Priority Class Name:	openshift-user-critical

valerymo commented:
> system-node-critical

`system-node-critical` - it's just for example , for testing. It's taken from ApiManager CR. 
If Priority Class Name not defined in CR -  it's not set in DC/Pod.

laurafitzgerald commented:
Thanks Valery. I understand it's used an example but just noting that PriorityClass is reserved for system critical pods so shouldn't be used for user workloads/applications.

valerymo commented:
> Thanks Valery. I understand it's used an example but just noting that PriorityClass is reserved for system critical pods so shouldn't be used for user workloads/applications.

Thank you Laura. Yes agree, also seems to me it's a Big Risk to have option set Default priority on application level for All pods together. Current approach seems to me - safe

valerymo commented:
APIManager CR example updated in validation section. Priority Class changed to `openshift-user-critical`
$ kubectl get deploymentconfigs -o name | xargs -I {} bash -c "echo -n '{} -> priorityClassName: '; kubectl get {} -o jsonpath='{.spec.template.spec.priorityClassName}' | yq e -P"
deploymentconfig.apps.openshift.io/apicast-production -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/apicast-staging -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/backend-cron -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/backend-listener -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/backend-redis -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/backend-worker -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-app -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-memcache -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-postgresql -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-redis -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-searchd -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/system-sidekiq -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/zync -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/zync-database -> priorityClassName: openshift-user-critical
deploymentconfig.apps.openshift.io/zync-que -> priorityClassName: openshift-user-critical

valerymo commented:
@laurafitzgerald , @eguzki could you please tell if any changes required in current solution.  
I checked `eng-long-lived` cluster. All DCs have `priorityClassName: rhoam-pod-priority`.  
Thank you

eguzki reviewed (approved):
Good Job @valerymo :medal_military:

valerymo commented:
@eguzki , @laurafitzgerald , Thank you very much for your kind help and review!
